File: examples/carp/zero-array-ic.c
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
//:: cases ZeroArrayIC-C
//:: tools silicon
//:: verdict Pass


/*@
  requires \length(ar)==len ** ArrayPerm(ar,0,1,len,1);
  ensures  \length(ar)==len ** ArrayPerm(ar,0,1,len,1);
  ensures  (\forall int k ; 0 <= k && k < len ; ar[k]==0 );
@*/
void zero_array(int ar[],int len){
    //@ assert \length(ar)==len ;
    for(int i=0;i < len;i++)
       /*@  requires \length(ar)==len ** Perm(ar[i],1);
            ensures  \length(ar)==len ** Perm(ar[i],1) ** ar[i]==0;
       @*/
    {
        ar[i]=0;
    }
}

File: examples/carp/ZeroArrayIC.java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
//:: cases ZeroArrayIC
//:: tools silicon
//:: verdict Pass

public class ZeroArray {
    /*@ 
        context_everywhere ar != null;
        requires Perm(ar[*],1);
        ensures Perm(ar[*],1) ** (\forall int k ; 0 <= k && k < ar.length ; ar[k]==0 ) ;
     */
    public void zero_array(int ar[]){
        //int N=ar.length;
        for(int i=0;i<ar.length;i++)
        /*@  requires Perm(ar[i],1);
             ensures  Perm(ar[i],1) ** ar[i]==0;
        @*/
        {
            ar[i]=0;
        }
    }
}

